### PR TITLE
Add Signed blocks

### PR DESCRIPTION
**Pull request by ChihChengLiang**
### What was wrong?

This is part of #1434, should be merged after #1471 

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] Clean up commit history

[//]: # (For important changes that should go into the release notes please add a newsfragment file as explained here: https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

[//]: # (See: https://trinity-client.readthedocs.io/en/latest/contributing.html#pull-requests)
- [ ] Add entry to the [release notes](https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

File: eth2/beacon/typing.py [hunk 1]
```diff
@@ -30,12 +28,6 @@ def __str__(self) -> str:
 
 DomainType = NewType("DomainType", bytes)  # bytes of length 4
 
-# Use HashTreeRoot and SigningRoot for Merkle tree root of SignedSerializable
-# Merkle tree root with signature field included
-HashTreeRoot = NewType("HashTreeRoot", Hash32)
-# Merkle tree root with signature field excluded
-SigningRoot = NewType("SigningRoot", Hash32)
```
Comment (ChihChengLiang) [inline eth2/beacon/typing.py]: Both `SigningRoot` and `HashTreeRoot` are gone. Use `eth_typing.Hash32` only.

File: eth2/beacon/state_machines/forks/serenity/blocks.py [hunk 2]
```diff
@@ -1,5 +1,9 @@
-from eth2.beacon.types.blocks import BeaconBlock
+from eth2.beacon.types.blocks import BeaconBlock, SignedBeaconBlock
 
 
 class SerenityBeaconBlock(BeaconBlock):
```
Comment (ChihChengLiang) [inline eth2/beacon/state_machines/forks/serenity/blocks.py]: Not sure if we want the unsigned one. But we can remove this in the next PR for better reviewability.

File: eth2/beacon/state_machines/forks/serenity/state_transitions.py [hunk 3]
```diff
@@ -17,19 +18,23 @@ def __init__(self, config: Eth2Config):
...
+        target_slot = signed_block.message.slot if signed_block else future_slot
         assert target_slot is not None
 
         state = process_slots(state, target_slot, self.config)
 
-        if block:
-            state = process_block(state, block, self.config, check_proposer_signature)
+        if signed_block:
+            if check_proposer_signature:
+                validate_proposer_signature(
```
Comment (ChihChengLiang) [inline eth2/beacon/state_machines/forks/serenity/state_transitions.py]: This check was moved from `process_block_header`. Mainly because the fixture tests assume `process_block_header` takes an unsigned header.

File: eth2/beacon/state_machines/forks/skeleton_lake/__init__.py [hunk 4]
```diff
@@ -14,6 +17,7 @@ class SkeletonLakeStateMachine(BeaconStateMachine):
     config = MINIMAL_SERENITY_CONFIG
 
     # classes
+    signed_block_class = SerenitySignedBeaconBlock
     block_class = SerenityBeaconBlock
```
Comment (ChihChengLiang) [inline eth2/beacon/state_machines/forks/skeleton_lake/__init__.py]: We are defining both `signed_block_class` and `block_class` for the time being

File: tests/eth2/core/beacon/state_machines/forks/test_serenity_block_proposer_slashing_validation.py [hunk 5]
```diff
@@ -35,10 +35,11 @@ def test_validate_proposer_slashing_epoch(genesis_state, keymap, config):
     # Valid
     validate_proposer_slashing_epoch(valid_proposer_slashing, config.SLOTS_PER_EPOCH)
 
-    header_1 = valid_proposer_slashing.header_1.set(
-        "slot", valid_proposer_slashing.header_2.slot + 2 * config.SLOTS_PER_EPOCH
+    invalid_proposer_slashing = valid_proposer_slashing.transform(
+        ("signed_header_1", "message", "slot"),
```
Comment (ChihChengLiang) [inline tests/eth2/core/beacon/state_machines/forks/test_serenity_block_proposer_slashing_validation.py]: `transform` API saves us from creating another `signed_header_1` and `message` objects.

File: tests/eth2/fixtures/test_epoch_processing.py [hunk 6]
```diff
@@ -33,6 +35,7 @@ def test_justification_and_finalization(test_case):
         },
     }
 )
+@pytest.mark.skip(reason="Need to fix queue rate")  # TODO: Fix in #1480
 def test_registry_updates(test_case):
```
Comment (ChihChengLiang) [inline tests/eth2/fixtures/test_epoch_processing.py]: Highlight: Skip fixture tests here. Plan to do it in another PR

File: tests/eth2/fixtures/test_sanity.py [hunk 7]
```diff
@@ -23,15 +23,17 @@ def test_slots(test_case):
...
     {
         "config_types": (Minimal,),
         "test_types": {SanityTestType: lambda handler: handler.name == "blocks"},
     }
 )
 def test_blocks(test_case):
+    # TODO: Remove when upgrading to 0.9.4.
+    # See https://github.com/ethereum/eth2.0-specs/pull/1544
+    if test_case.name in ("invalid_state_root", "zero_block_sig", "invalid_block_sig"):
+        pytest.skip("Missing pre state in this case")
```
Comment (ChihChengLiang) [inline tests/eth2/fixtures/test_sanity.py]: Highlight: Skip fixture tests here. Need to add these back when we upgrade the fixture to 0.9.4.

File: tests/libp2p/bcc/test_syncing.py [hunk 8]
```diff
@@ -86,11 +86,13 @@ def assert_synced(alice, bob, correct_branch):
 
 @pytest.mark.asyncio
 async def test_sync_from_old_head(request, event_loop, event_bus, genesis_state):
-    genesis_block = get_genesis_block(genesis_state.hash_tree_root, BeaconBlock)
-    alice_branch = (genesis_block,) + BeaconBlockFactory.create_branch(
+    genesis_block = SignedBeaconBlockFactory(
+        message__state_root=genesis_state.hash_tree_root
```
Comment (ChihChengLiang) [inline tests/libp2p/bcc/test_syncing.py]: Not sure if frequently using factories with the style like `message__state_root` sounds too crazy.

File: tests/eth2/fixtures/test_epoch_processing.py (hunk 6 above)
Comment (ralexstokes) [reply tests/eth2/fixtures/test_epoch_processing.py]: make a tracking issue if you think you will forget!! :)

File: tests/eth2/fixtures/test_sanity.py (hunk 7 above)
Comment (ralexstokes) [reply tests/eth2/fixtures/test_sanity.py]: ditto on tracking issue

File: tests/libp2p/bcc/test_syncing.py (hunk 8 above)
Comment (ralexstokes) [reply tests/libp2p/bcc/test_syncing.py]: it's a little wonky bc it reaches into the internals of the factory library *but* it seems like this is the most direct way to do this in cases like these, so i wouldn't say we need to worry too much about it for now

File: .circleci/get_eth2_fixtures.sh [hunk 9]
```diff
@@ -1,9 +1,9 @@
 #!/usr/bin/env bash
 
 if [ ! -d "./eth2-fixtures/tests" ]; then
-  wget -c https://github.com/ethereum/eth2.0-spec-tests/releases/download/v0.9.2/general.tar.gz
-  wget -c https://github.com/ethereum/eth2.0-spec-tests/releases/download/v0.9.2/mainnet.tar.gz
-  wget -c https://github.com/ethereum/eth2.0-spec-tests/releases/download/v0.9.2/minimal.tar.gz
+  wget -c https://github.com/ethereum/eth2.0-spec-tests/releases/download/v0.9.3/general.tar.gz
```
Comment (ralexstokes) [inline .circleci/get_eth2_fixtures.sh]: yay!

File: eth2/beacon/chains/base.py [hunk 10]
```diff
@@ -60,7 +61,7 @@ def from_genesis(
         cls,
         base_db: AtomicDatabaseAPI,
         genesis_state: BeaconState,
-        genesis_block: BaseBeaconBlock,
+        genesis_block: BaseSignedBeaconBlock,
```
Comment (ralexstokes) [inline eth2/beacon/chains/base.py]: looks like the genesis block is not signed.

File: eth2/beacon/chains/base.py [hunk 11]
```diff
@@ -107,7 +108,7 @@ def on_attestation(self, attestation: Attestation, slot: Slot = None) -> None:
     # State API
     #
     @abstractmethod
-    def get_state_by_root(self, root: HashTreeRoot) -> BeaconState:
+    def get_state_by_root(self, root: Hash32) -> BeaconState:
```
Comment (ralexstokes) [inline eth2/beacon/chains/base.py]: what's the rationale for removing this type? looks like the spec keeps a `Root` type. i'm not opposed to using `Hash32` here but just curious what the thinking is...

File: eth2/beacon/scripts/quickstart_state/generate_beacon_genesis.py [hunk 12]
```diff
@@ -52,7 +52,7 @@ def _main():
 
     genesis_time = int(time.time())
     print(f"creating genesis at time {genesis_time}")
-    genesis_state = state.copy(genesis_time=genesis_time)
+    genesis_state = state.set("genesis_time", genesis_time)
```
Comment (ralexstokes) [inline eth2/beacon/scripts/quickstart_state/generate_beacon_genesis.py]: nice catch!

File: eth2/beacon/state_machines/forks/serenity/block_validation.py [hunk 13]
```diff
@@ -80,9 +81,11 @@ def validate_proposer_is_not_slashed(
 
 
 def validate_proposer_signature(
-    state: BeaconState, block: BaseBeaconBlock, committee_config: CommitteeConfig
+    state: BeaconState,
+    signed_block: BaseSignedBeaconBlock,
+    committee_config: CommitteeConfig,
 ) -> None:
-    message_hash = block.signing_root
+    message_hash = signed_block.message.hash_tree_root
```
Comment (ralexstokes) [inline eth2/beacon/state_machines/forks/serenity/block_validation.py]: it looks like the correct `message_hash` is the output of `compute_signing_root(signed_block.message, get_domain(state, DOMAIN_BEACON_PROPOSER))` -- not sure if this update comes in a later PR or if I'm on the wrong spec version -- just mentioning it now

File: eth2/beacon/state_machines/forks/serenity/block_validation.py [hunk 14]
```diff
@@ -199,20 +207,20 @@ def validate_proposer_slashing_is_slashable(
...
-    header: BeaconBlockHeader,
+    header: SignedBeaconBlockHeader,
     pubkey: BLSPubkey,
     slots_per_epoch: int,
 ) -> None:
     try:
         bls.validate(
             pubkey=pubkey,
-            message_hash=header.signing_root,
+            message_hash=header.message.hash_tree_root,
```
Comment (ralexstokes) [inline eth2/beacon/state_machines/forks/serenity/block_validation.py]: ditto

File: eth2/beacon/tools/builder/proposer.py [hunk 15]
```diff
@@ -99,26 +103,26 @@ def create_block_on_state(
...
-    state, block = state_machine.import_block(
-        block, state, check_proposer_signature=False
+    state, signed_block = state_machine.import_block(
+        signed_block, state, check_proposer_signature=False
     )
 
     # Sign
     signature = sign_transaction(
-        message_hash=block.signing_root,
+        message_hash=signed_block.message.hash_tree_root,
```
Comment (ralexstokes) [inline eth2/beacon/tools/builder/proposer.py]: ditto on message here

File: eth2/beacon/tools/builder/validator.py [hunk 16]
```diff
@@ -239,24 +238,26 @@ def create_block_header_with_signature(
...
+) -> SignedBeaconBlockHeader:
     block_header = BeaconBlockHeader.create(
         slot=state.slot,
         parent_root=parent_root,
         state_root=state_root,
         body_root=body_root,
     )
     block_header_signature = sign_transaction(
-        message_hash=block_header.signing_root,
+        message_hash=block_header.hash_tree_root,
```
Comment (ralexstokes) [inline eth2/beacon/tools/builder/validator.py]: ditto on signing

File: eth2/beacon/types/blocks.py [hunk 17]
```diff
@@ -189,51 +177,106 @@ def from_root(
...
+    ) -> TSignedBeaconBlock:
+        return super().create(message=message, signature=signature)
+
+    @property
+    def parent_root(self) -> Hash32:
+        return self.message.parent_root
+
+    @property
+    def signing_root(self) -> Hash32:
+        # TODO: Remove this soon too
```
Comment (ralexstokes) [inline eth2/beacon/types/blocks.py]: what is blocking removing this now?

File: tests/components/eth2/apis/test_http.py [hunk 18]
```diff
@@ -52,7 +52,11 @@ def ipc_path():
         )
 
         chaindb.persist_state(genesis_state)
-        chaindb.persist_block(genesis_block, genesis_block.__class__, fork_choice_scoring)
+        chaindb.persist_block(
+            SignedBeaconBlock.create(message=genesis_block),
+            SignedBeaconBlock,
```
Comment (ralexstokes) [inline tests/components/eth2/apis/test_http.py]: should this be a particular fork's `SignedBeaconBlock`... in this case it would be a `SerenitySignedBeaconBlock`, correct?

Comment (ralexstokes) [reply tests/components/eth2/apis/test_http.py]: i guess the generic thing here would be that a `BeaconBlock` class can produce its signed equivalent, like `SerenityBeaconBlock.signed_class` yields `SerenitySignedBeaconBlock` -- although this gets close to introducing circularity in the classes which we don't want....

File: tests/eth2/fixtures/test_sanity.py (hunk 7 above)
Comment (ChihChengLiang) [reply tests/eth2/fixtures/test_sanity.py]: Created issue #1483

File: tests/eth2/fixtures/test_epoch_processing.py (hunk 6 above)
Comment (ChihChengLiang) [reply tests/eth2/fixtures/test_epoch_processing.py]: Created the tracking issue here #1480. Thank you for reminding me.

File: eth2/beacon/chains/base.py (hunk 10 above)
Comment (ChihChengLiang) [reply eth2/beacon/chains/base.py]: Since the chaindb takes signed blocks, we converted genesis block to a signed block. But on second thought, we should do the conversion inside the chaindb's API instead of do it outside. I'll fix that.

File: eth2/beacon/chains/base.py (hunk 11 above)
Comment (ChihChengLiang) [reply eth2/beacon/chains/base.py]: `HashTreeRoot` was originally introduced to differentiate with the old `SigningRoot` and was a subtype of `Hash32`. But you are right, we should use a `Root` type to differentiate with `Byte32` type data like `withdrawal_credentials`, eth1 `block_hash`s, and `randao_mixes`, etc.

File: eth2/beacon/state_machines/forks/serenity/block_validation.py (hunk 13 above)
Comment (ChihChengLiang) [reply eth2/beacon/state_machines/forks/serenity/block_validation.py]: `compute_signing_root` was introduced in sepc version 0.10.0. So we are not there yet.

File: eth2/beacon/types/blocks.py (hunk 17 above)
Comment (ChihChengLiang) [reply eth2/beacon/types/blocks.py]: Made this API to prevent introducing more diffs in this PR, it lets some parts of the old references to `signing_root` still work.

Tracking issue in #1480, will fix in the following PR.

**Comment by ChihChengLiang** [comment]
@hwwhww Per https://github.com/ethereum/trinity/pull/1434#issuecomment-572001539, we don't need to rename `get_genesis_block_root` to `get_signed_genesis_block_root`. Whether the genesis_block signed or not, the function should return the same `Root`, the root of the message.

File: tests/components/eth2/apis/test_http.py (hunk 18 above)
Comment (ChihChengLiang) [reply tests/components/eth2/apis/test_http.py]: > "should this be a particular fork's SignedBeaconBlock?"

Yes, it should be. But in tests, maybe using the right block class doesn't matter that much.

**Comment by ChihChengLiang** [comment]
Merge to unblock other tasks. Post PR feedbacks are welcomed. Thank you Alex for the review.
